### PR TITLE
refactor!: migrate to typescript

### PR DESCRIPTION
**Pull request by rnike**
continue #95

Resolve todos and fix some issues, ~~also changed the compiler to `tsc` because this is how I usually do, `@react-native-community/bob` was removed.~~

This is a huge change, before public release, I think we should have a beta version first

### CODE REVIEW

**Comment by rnike** [comment]
the conflict has been resolved

**Comment by mikehardy** [comment]
Ironically I just spent most of my "open source" time today reviewing another module's typescript re-write, so I don't have time just this moment - but similar to that one I cannot overstate how happy I am to see this, and how happy I will be to have it in typescript. So thank you

One quick question - the switch from bob to tsc. using bob let's us easily generate mutliple derivative JS targets, commonjs and modules, at the same time. So is kind of useful. Do you feel strongly about raw tsc, or even if bob is unfamiliar to you, would it be okay to leave it and preserve our ability to intermediate between our typescript source and various/possibly-multiple JS transpiles?

**Comment by rnike** [comment]
@mikehardy The main point I remove bob is I didn't see why we need multiple kinds of js output, and I kind of want to reduce the package size on npm (including not uploading the src folder) to just make it neat, but maybe the whole thing (bob removal and size reduce) should be fully discussed before I go further, I'm a little bit too fast on this without consideration, let me put bob back on this PR, thanks for the reminder.

File: src/FBGraphRequestManager.ts
Comment (rnike) [inline src/FBGraphRequestManager.ts:39]: `(param as any)` causes the warning, but I currently cannot figure out the solution

**Comment by mikehardy** [comment]
Just about including source or not: uploading src is a strict requirement for me. The majority of pull requests come from people that just reach into node_modules and just fix something, the PR is an afterthought. If you remove the src and build machinery you lose a huge percentage of contributions in my experience. I do not see uploading a few extra text files in the npmjs pack binary as being any sort of problem, compared to the general problem that in open source, people that actually contribute are far too infrequent...

**Comment by rnike** [comment]
@mikehardy agree with you uploading a few extra text files isn't a big deal, I've added bob back and this PR I think is ready to merge if no further issue.

**Comment by mikehardy** [comment]
great - it's the weekend here in the holidays so review might be a bit delayed but please don't take that as any signal - I'm really excited to get this done. Thanks again for posting it

File: package.json
Comment (rnike) [inline package.json:8]: Probably will lead to an issue, not everyone is using typescript, I should remove this

**Comment by rnike** [comment]
It will be wonderful if there's a beta version on npm to test it properly, currently I am just pushing it to my repo and yarn add from there.

None is it necessary, just consider it as a `nice to have`

File: src/FBAccessToken.ts
Comment (rnike) [inline src/FBAccessToken.ts:149]: check should it be `accessToken?: FBAccessToken`

File: src/FBAppEventsLogger.ts
Comment (rnike) [inline src/FBAppEventsLogger.ts:150]: not sure `: void` is necessary, remove it if is not

Comment (rnike) [reply src/FBAppEventsLogger.ts]: unnecessary type declarations have been removed

File: src/FBAccessToken.ts
Comment (rnike) [reply src/FBAccessToken.ts]: it is surely `FBAccessToken | null` according to the native implementation

File: src/FBGraphRequest.ts
Comment (rnike) [inline src/FBGraphRequest.ts:64]: according to the constructor, `callback` will never be undefined,  so does `config`

File: src/models/FBShareOpenGraphValueContainer.ts
Comment (rnike) [inline src/models/FBShareOpenGraphValueContainer.ts:42]: this should be `properties: OpenGraphProperties` according to previous type

File: src/FBProfile.ts
Comment (rnike) [inline src/FBProfile.ts:8]: After looked into the source code of both `facebook-ios-sdk` and `facebook-android-sdk`, I think the correct type of each key is  `string | undefined | null`

Comment (rnike) [reply src/FBProfile.ts:8]: fixed

**Review by rnike — COMMENTED** [review]
self reviewed all files, issues found are all solved

**Comment by mikehardy** [comment]
> It will be wonderful if there's a beta version on npm to test it properly, currently I am just pushing it to my repo and yarn add from there.
> 
> None is it necessary, just consider it as a `nice to have`

A strict requirement of the migration for me is that it is buildable from source, from the installed version or from a version that is in git (so people may use a repo or branch or commit ref) then they should be able to change into the node_modules/react-native-fbsdk-next directory and run `yarn` (or `yarn && yarn build` or whatever) and it works. If that's okay then it's okay to alpha test at least, you can put out a call for testers on the roadmap issue perhaps? I can test it as well. I'm still behind deadline on one project and now it's the holidays so I haven't had time personally but this is critical for the repository's future I think, so it *will* get reviewed and merged soon I hope :-), thanks for your patience

**Comment by rnike** [comment]
I've created a pre-release named `v7.0.0-alpha.1` including the recent feature I've made in my repo [rnike/react-native-fbsdk#v7.0.0-alpha.1](https://github.com/rnike/react-native-fbsdk-next/releases/tag/v7.0.0-alpha.1), but what I really want to do is to publish `v7.0.0-alpha.1` to npm for a proper testing, @mikehardy any idea how can I achieve that?

**Comment by mikehardy** [comment]
I can try after I review it. The holidays mean I'm quite short on time though, apologies

**Comment by rnike** [comment]
@mikehardy That's ok, this is no hurry. Thank you for still replying during your holidays.

**Comment by mikehardy** [comment]
status for me: 
1) I've pulled the branch and run the example app successfully, as a basic criteria for merge - everything is :100: 
2) I want to review the actual changes but may just not have time - so the strategy I will actually use is exactly as @rnike suggests - I'll see about getting a beta version out, then people may test integrate it. We can put a time box around how long we wait prior to full release perhaps? Maybe at the end of the first week of January? In the end people do not *have* to upgrade and we'll mark it as a major version just in case so it should be okay either way

**Review by mikehardy — APPROVED** [review]
Merging this into a new beta branch, and semantic-release should auto-magically publish that in to a beta tag on npmjs.com

**Comment by mikehardy** [comment]
I'm having a really frustrating time getting the version and tag lined up the way I like.
I have so far been unable to trigger a forced breaking change, and I think I know why but I am out of time for now on holidays to fix it.

The release is officially out there on the `beta` tag on npmjs.com so anyone may install it. It is still in the v6 series though, which is unwanted. It is released from the beta branch here until merged back to master. beta branch is strictly ahead of master at the moment so that part is easy to manage.

Future steps:
- add configuration to semantic release on beta branch so it understands the `!` syntax in commit messages for breaking changes (I did this in some other repo, I think react-native-google-ads so can copy from there)
- make another `!` commit on beta branch and trigger release action on beta branch so beta tag on npmjs.com lines up with a v7 beta release
- test integrate beta package with various work projects and my react-native-authentication-example-template to make sure it's good
- let it sit until january then merge to master and release there ?

**Comment by mikehardy** [comment]
This is now successfully out as v7.0.0-beta.1 on npmjs.com and on the beta branch, got it all working.
Now it can incubate - and this is a call for testers, any interested parties should test this in their projects and report problems before end of first week of January!

**Comment by rnike** [comment]
Thanks

**Pull request merged**
